src/mongo/mongoConnectionStrings.ts

```diff
@@ -49,6 +49,23 @@ export async function parseMongoConnectionString(connectionString: string): Prom
     let host: string;
     let port: string;
 
+    try {
+        // Attempt to connect to determine if the username/password need to be encoded
+        await connectToMongoClient(connectionString, appendExtensionUserAgent());
+    } catch (error) {
+        const parsedError: IParsedError = parseError(error);
+        if (parsedError.message.match(/unescaped/i)) {
+            const matches: RegExpMatchArray | null = connectionString.match(/^(.*):\/\/(.*):(.*)@(.*)/);
```

Comment:
How about pulling some of this stuff into it's own function and adding some unit tests? Other functions in this file already have it.